test/CMakeLists.txt

```diff
@@ -179,13 +197,24 @@ macro(try_add_performance_test)
     ${rmw_implementation}
     "ROS2"
     ${rmw_implementation}
+    "async"
   )
+
   if(COMM_TYPE_${rmw_implementation})
     add_performance_test(
       ${COMM_TYPE_${rmw_implementation}}
       ${COMM_TYPE_${rmw_implementation}}
       ${rmw_implementation}
-    )
+      "async"
+    )
+    if(${rmw_implementation} STREQUAL "rmw_fastrtps_cpp")
```

Comment:
This adds a standalone test for Fast-RTPS, but doesn't add a "sync" test for `rmw_fastrtps_cpp`. The `RMW_FASTRTPS_USE_QOS_FROM_XML` environment variable affects the ROS2 test, not the standalone one here.

So there should definitely be a similar conditional `add_performace_test()` above with the `ROS2` one, but that leads to the question, whether stand-alone Fast-RTPS can be configured similarly to the RMW?

Reply:
Right now, there is no way to configure async with any of the `rmw_*_cpp`. You have a look to this line in the code 

https://github.com/ros2/performance_test/blob/run_on_buildfarm/performance_test/src/experiment_configuration/experiment_configuration.cpp#L226

But the standalone FastRTPS allows us to configured async using these two env variables:
 - `RMW_FASTRTPS_USE_QOS_FROM_XML`
 - `FASTRTPS_DEFAULT_PROFILES_FILE`

Reply:
If I may pitch in again, perfomance_test does no allow to disabling async mode for ROS2 communication mean. However, that does not mean that the communication is going to be asynchronous (in fact there is not such a thing a CycloneDDS async, so it will always be sync both in standalone and ROS2 no matter what).

As for rmw_fastrtps_cpp, even though performance_test cannot disable async mode, it still can be disabled with `RMW_FASTRTPS_USE_QOS_FROM_XML` and `FASTRTPS_DEFAULT_PROFILES_FILE`.

For sync publishing on standalone FastRTPS, there is no need of those env, since in that case performance_test allows for `--disable_async`.

Reply:
Now it's clear for me. Thank you. I will fix it

Reply:
I'm trying this appoach but I'm getting some error when I'm trying to run perf_test with fastrtps async. I configured FASTRTPS_DEFAULT_PROFILES_FILE and RMW_FASTRTPS_USE_QOS_FROM_XML. Then I run 

```
ros2 run performance_test perf_test -c ROS2 -t Array1k --max_runtime 30 --roundtrip_mode Main
...
terminate called after throwing an instance of 'eprosima::fastcdr::exception::NotEnoughMemoryException'
  what():  Not enough memory in the buffer stream
```

same if I use this other approach:

```
ros2 run performance_test perf_test -c ROS2 -t Array1k --max_runtime 30 --num_sub_threads 0 --num_pub_threads 1
....
terminate called after throwing an instance of 'eprosima::fastcdr::exception::NotEnoughMemoryException'
  what():  Not enough memory in the buffer stream
```

Reply:
When setting RMW_FASTRTPS_USE_QOS_FROM_XML=1, the publication mode and history memory policy will take whatever set on the XML. If nothing is set there, they will take Fast-RTPS default values.

| Attribute | Default ROS 2 | Default Fast-RTPS |
|-|-|-|
| publication mode | async | sync |
| history memory policy | PREALLOCATED_WITH_REALLOC | PREALLOCATED |

In your XML you don't set the history memory policy, so it's being set to PREALLOCATED. If there is any unbounded type, you will get allocation memory issues. You can try to put a publisher and subscriber profile in your XML to tackle this:

```xml
    <profiles>
        <publisher profile_name="publisher profile" is_default_profile="true">
            <qos>
                <publishMode>
                    <kind>SYNCHRONOUS</kind>
                </publishMode>
            </qos>
            <historyMemoryPolicy>PREALLOCATED_WITH_REALLOC</historyMemoryPolicy>
        </publisher>

        <subscriber profile_name="subscriber profile" is_default_profile="true">
            <historyMemoryPolicy>PREALLOCATED_WITH_REALLOC</historyMemoryPolicy>
        </subscriber>
    </profiles>
```

Reply:
If ros2/performance_test#6 unblocks getting a synchronous rmw_fastrtps_cpp test online, let's get that added here.

Reply:
Do you mean as a comment?

Reply:
> rmw_fastrpts_cpp: Async by default. Can be sync with > RMW_FASTRTPS_USE_QOS_FROM_XML and FASTRTPS_DEFAULT_PROFILES_FILE.

Does that work? If so, why not have the test?

Reply:
@cottsay @ahcorde . Both `rmw_fastrtps_cpp` and `rmw_fastrtps_dynamic_cpp` can be configured both sync and async. I think both should be in the plots.

Actually, I think it makes more sense to have separate plots for sync and async, instead of all mixed up, or at least specify sync or async in all the series, cause right now it is only specified for Fast-RTPS, and people are probably unaware of the default publication modes for each implementation. If the plots are indeed split in 2, then some of the implementation would be present in one but not the other.